1320 file index

jo-pol opened the pull request:
#Fixes EASY-1320

#### When applied it will
* create solr documents for items in files.xml of bags
* for a next PR:
  * complex spatial coverage types
  * implementation as a service

# please squash

#### Where should the reviewer @DANS-KNAW/easy start?

The `solrLiterals` of the classes `DDM`, `Bag` and `FileItem` show what is indexed along with the file content. The ApplicationWiring class controls the commits (once per per bag, whatever succeeded) and what is logged (only problems and what was committed anyway, no progress).


#### How should this be manually tested?

* deploy the related pull request on `deasy`
* adjust configuration to use the bag-store of **a**easy to try proper pdbs bags
* deploy and run _on_ `deasy`
  or use the run scripts of [dev-tools] to run _with_ deasy

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-dtap                      | [#131](https://github.com/DANS-KNAW/easy-dtap/pull/131)     | required for manual testing
easy-bag-store  | todo | return the proper content type, required for implementation on production. Nice to have: file size in files.xml would eliminate a header request
easy-build | todo | add the project

jo-pol committed:
API

jo-pol committed:
missed the example in the README

jo-pol committed:
Reading store/file names. API: optional store for init, no more add.

jo-pol committed:
working API

jo-pol committed:
todo too much

jo-pol committed:
manage resources

jo-pol committed:
boiler plate of the App class

jo-pol committed:
moved logic from CommandLineOptions to vault class

jo-pol committed:
dropped obsolete method

jo-pol committed:
unit tests

jo-pol committed:
refactoring

jo-pol committed:
collect required info

jo-pol committed:
class -> trait + formatting

jo-pol committed:
back to class

jo-pol committed:
fixed pure expression

jo-pol committed:
rename + formatting

jo-pol committed:
moved DebugEnhancedLogging

jo-pol committed:
extracted VaultIO class

jo-pol committed:
more tests

jo-pol committed:
building a solr document

jo-pol committed:
license headers

jo-pol committed:
improved tests

jo-pol committed:
managed to add a document

jo-pol committed:
fixed delete, transaction per bag, success/failure rate in feedback message

jo-pol committed:
using streams

jo-pol committed:
refactoring solr literal creation

jo-pol committed:
refactored FileItem construction

jo-pol committed:
refactored FileItem constructor

jo-pol committed:
Windows-proof unit tests, still need to mock commit

jo-pol committed:
Windows-proof unit tests, offline testing

jo-pol committed:
retries of updates without streams

jo-pol committed:
new DDM test (fails)

jo-pol committed:
fixed DDM test

jo-pol committed:
test for file items

jo-pol committed:
feedback message with just the number of retries

jo-pol committed:
vocabularies (quick and dirty)

jo-pol committed:
another todo

jo-pol committed:
added canConnect to test

jo-pol committed:
count mixed results

jo-pol committed:
abr tables no longer merged

jo-pol committed:
todo was fixed some commits ago

jo-pol committed:
multiple cretor types and instances, other types of relations

see https://github.com/DANS-KNAW/easy-schema/blob/acb65062acf17cb43d92264c38fd6aadd63529d7/src/main/assembly/dist/docs/examples/ddm/example2.xml#L70

jo-pol committed:
multiple cretor types and instances, other types of relations

see https://github.com/DANS-KNAW/easy-schema/blob/acb65062acf17cb43d92264c38fd6aadd63529d7/src/main/assembly/dist/docs/examples/ddm/example2.xml#L70

jo-pol committed:
"contain only" gave false positives

jo-pol committed:
deal with prefixed attributes

jo-pol committed:
formatting

jo-pol committed:
retry on exception or non-zero status

jo-pol committed:
don't index accessRights == NONE

jo-pol committed:
refactored error handling

jo-pol committed:
refactored error handling

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,88 @@
+
+  def initAllStores(): Try[FeedBackMessage] = {
+    getStoreNames
+      .flatMap(storeNames => updateStores(storeNames))
```
`.flatMap(updateStores)`

jo-pol committed:
obsolete dependency

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,88 @@
+
+  def initSingleStore(storeName: String): Try[FeedBackMessage] = {
+    getBagIds(storeName)
+      .flatMap(bagIds => updateBags(storeName, bagIds))
```
either

`.flatMap(updateBags(storeName, _))`

or curry `updateBags` and do

`.flatMap(updateBags(storeName))`

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,88 @@
+  private def updateStores(storeNames: Seq[String]): Try[Seq[FeedBackMessage]] = {
+    storeNames
+      .map(initSingleStore)
+      .collectResults.recoverWith {
```
`recoverWith` on newline, then the formatting on the next lines will be better...

same for similar instances below

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,88 @@
+    storeNames
+      .map(initSingleStore)
+      .collectResults.recoverWith {
+      case t: CompositeException => throw new Exception(s"Tried to update ${ storeNames.size } stores, ${ t.getMessage() }", t)
```
Make the exceptions as classes with nice names, e.g. `UpdateStoreException`.

```scala
case class UpdateStoreException(n: Int, msg: String, t: Throwable) extends Exception(s"Tried to update $n stores, $msg", t)
```

same for similar instances below

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,88 @@
+
+  private def collectMixedResults(bagId: String, results: Seq[Try[FeedBackMessage]]) = {
+    val nrOfFailures = results.count(_.isFailure)
+    val successMessages = results.filter(_.isSuccess).map(_.get)
```
you could use `results.collect { case Success(msg) => msg }` here instead

jo-pol committed:
can't mock SolrClient

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,88 @@
+    collectMixedResults(bag.bagId, results)
+  }
+
+  private def collectMixedResults(bagId: String, results: Seq[Try[FeedBackMessage]]) = {
```
after reading this method more carefully, a couple of things come to mind:

1. why don't you inline this method with `createDocs`? It's only used there, after all...
2. I'd use a different implementation:
    a. first do a `results.collectResults`, which has type `Try[Seq[FeedBackMessage]]`
    b. if this results in a `Success`, you know that the number of `Failure`s in `results` is zero. The `successMessages` can then directly be found inside the `Success`
    c. if the result is a `Failure`, you get a `Failure(CompositeExceptions(listOfExceptions))`. You can either leave this as is, or wrap it in another `Exception` as you do now.

jo-pol replied:
1: I extracted it to stress the similarity with the normal collectResult. Moved it to an implicit class.
2c: still might have to collect some actions that did succeed and count (as in 2b) how many were indexed with or without content. But I could drop one loop.

rvanheest replied:
You may also be interested in a `seq.partition(...)` function?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/Command.scala:
```diff
@@ -43,7 +42,16 @@ object Command extends App with DebugEnhancedLogging {
     .doIfFailure { case e => logger.error(e.getMessage, e) }
     .doIfFailure { case NonFatal(e) => println(s"FAILED: ${ e.getMessage }") }
 
+
```
empty line?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/Command.scala:
```diff
@@ -21,10 +21,9 @@ import resource._
-  type FeedBackMessage = String
 
   val configuration = Configuration()
   val commandLine: CommandLineOptions = new CommandLineOptions(args, configuration) {
```
Is there any need not to do the `verify()` below inside the `CommandLineOptions`? Not sure about this. I need a reminder if you really need it this way, because I did it inside `CommandLineOptions` for most of the EASY-1291 issues.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/Command.scala:
```diff
@@ -43,7 +42,16 @@ object Command extends App with DebugEnhancedLogging {
+        case update @ commandLine.update => app.update(update.bagStore(), update.bagUuid())
+        case delete @ commandLine.delete => app.delete(delete.bagUuid())
+        case init @ commandLine.init if init.bagStore.isSupplied => app.initSingleStore(init.bagStore())
+        case init @ commandLine.init if !init.bagStore.isSupplied => app.initAllStores()
```
you don't need the extra `!init.bagStore.isSupplied` here (can you figure out why?).

alternatively you could do

```scala
case init @ commandLine.init => init.bagStore.toOption
    .map(app.initSingleStore)
    .getOrElse(app.initAllStores())
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/CommandLineOptions.scala:
```diff
@@ -15,18 +15,23 @@
   appendDefaultToDescription = true
   editBuilder(_.setHelpWidth(110))
   printedName = "easy-update-solr4files-index"
   private val SUBCOMMAND_SEPARATOR = "---\n"
```
You forgot `version(configuration.version)` above this line.

jo-pol replied:
Did I forgot or did the archetype forgot?

rvanheest replied:
probably archetype forgot to add it. Maybe you could do a PR there?

rvanheest reviewed (changes requested):
Here is a first batch of comments.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Bag.scala:
```diff
@@ -0,0 +1,62 @@
+    val key = "EASY-User-Account"
+    vaultIO.linesFrom(storeName, bagId, "bag-info.txt")
+      .filter(_.trim.startsWith(key))
+      .map(_.trim.replace(key, "").trim.replace(":", "").trim)
```
is all this trimming really necessary?
I'd say you do a `.map(_.trim)` before the `filter` line: that way you don't need to do this `trim` twice.
Then using `collectFirst` you can do this easier than with this `map(...).head` construct.

jo-pol replied:
Or can I load this file like a `.properties` file? The ':' in stead of '=' pushed me in the DIY solution.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Bag.scala:
```diff
@@ -0,0 +1,62 @@
+
+  private val fileShas: Try[FileToShaMap] = Try {
+    vaultIO.linesFrom(storeName, bagId, "manifest-sha1.txt").map { line: String =>
+      val xs = line.trim.split("""\s+""")
```
do a pattern match on this array: `val Array(first, second) = line.trim.split("\s+", 2)`

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Bag.scala:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  def sha(path: String): String = {
+    fileShas.map(_.getOrElse(path, "")).getOrElse("")
```
you may prefer a `fold` construct here: `fileShas.fold("")(_.getOrElse(path, ""))`.
just a suggestion, if you don't like it, ignore it...

jo-pol replied:
Better create an empty `FileToShaMap` in case of problems than live with a tried one and ignore the failure anyway.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+
+class DDM(xml: Node) extends DebugEnhancedLogging {
+
+  val accessRights: String = (xml \ "profile" \ "accessRights").text
```
given that you do a lot of `xml \ "profile"` and `xml \ "dcmiMetadata"`, why don't you put them in values first?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+
+  val accessRights: String = (xml \ "profile" \ "accessRights").text
+  val solrLiterals: SolrLiterals =
+    (xml \ "dcmiMetadata" \ "identifier").filter(isDOI).map(n => ("dataset_doi", n.text)) ++
```
`withFilter` is more efficient here

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+      )) ++
+      (xml \ "dcmiMetadata" \ "subject").flatMap { n =>
+        val abrMap = getAbrMap(n) // FIXME
+        if (abrMap.isEmpty) Seq(("dataset_subject", n.text))
```
use a pattern match instead

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+        }
+      } ++
+      (xml \ "dcmiMetadata" \ "relation").map(n => ("dataset_relation", n.text))
+  (xml \ "dcmiMetadata" \ "conformsTo").map(n => ("dataset_relation", n.text))
```
@jo-pol do these work with both `ddm:conformsTo` and `dcterms:conformsTo`. Similar for `ddm:relation` vs. `dc:relation`

jo-pol replied:
the unit test shows it works

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+        }
+      } ++
+      (xml \ "dcmiMetadata" \ "relation").map(n => ("dataset_relation", n.text))
+  (xml \ "dcmiMetadata" \ "conformsTo").map(n => ("dataset_relation", n.text))
```
@PaulBoon are you sure you want all these subtypes of relations be marked as `dataset_relation`? You don't need these subtypes?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+        }
+      } ++
+      (xml \ "dcmiMetadata" \ "relation").map(n => ("dataset_relation", n.text))
+  (xml \ "dcmiMetadata" \ "conformsTo").map(n => ("dataset_relation", n.text))
```
@PaulBoon the new `ddm:<subrelation>` has a attribute `href`. You don't want to store them?

rvanheest replied:
@jo-pol similar for `ddm:relation` which can have a streaming surrogate relation... I guess you don't want to store this value?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+      .attribute(xsiURI, "type")
+      .map(_.text)
+      .filter(_.startsWith(abrPrefix))
+      .flatMap(abrMaps.get)
```
do you really need this `get` here. That's unsafe!!!

jo-pol replied:
`get` returns an option which I flatMap with the option returned by `attribute`, so what is unsafe?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+  private def loadVocabularies(xsdURL: String): Map[String, Map[String, String]] = {
+    val xmlDoc = resource.managed(
+      new URL(xsdURL).openStream()
+    ).acquireAndGet(XML.load) // TODO error handling
```
for error handling, use `map(...).tried` instead of `acquireAndGet`

jo-pol replied:
have to replace the IO anyway

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+  private def findKeyValuePairs(table: Node) = {
+    (table \\ "enumeration")
+      .map { node =>
+        val key: String = node.attribute("value").map(_.text).getOrElse("")
```
you may be interested in that `fold` construct here again?

rvanheest replied:
Also, we already know the type of this value, no need to state it explicitly

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,114 @@
+      .map { node =>
+        val key: String = node.attribute("value").map(_.text).getOrElse("")
+        val value = (node \ "annotation" \ "documentation").text
+        key -> value.replaceAll("\\s+", " ").trim
```
you do a lot of `trim`s, which is admirable! However, have you checked that you don't do to many trims. That's all extra computational complexity...

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/FileItem.scala:
```diff
@@ -0,0 +1,49 @@
+    // @formatter:off
+  }
+  val accessRights: String = ( xml \ "accessRights").map(_.text.trim).mkString
+  val path: String = xml.attribute("filepath")
```
`fold`?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/FileItem.scala:
```diff
@@ -0,0 +1,49 @@
+    case _                                  => "NONE"
+    // @formatter:off
+  }
+  val accessRights: String = ( xml \ "accessRights").map(_.text.trim).mkString
```
You're sure about this `mkString`? `["abc", "def", "ghi"].mkString == "abcdefghi"`

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,99 @@
+import scala.util.Try
+
+trait Solr {
+  this: DebugEnhancedLogging =>
```
use extends here

jo-pol replied:
Why? its done like that in other situations.

rvanheest replied:
when you compose the traits, you need to add DebugEnhancedLogging there as well. Also, this is not something you want to swap out for other (sub)types of logging. That's what @janvanmansum and I agreed upon.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,99 @@
+      addContentStream(stream)
+      setParam("literal.id", solrDocId)
+      setParam("literal.easy_file_size", stream.getSize.toString)
+      solrLiterals.foreach { case (key, value) if value.trim.nonEmpty =>
```
```scala
for ((key, value) <- solrLiterals;
    if value.trim.nonEmpty) {
  ...
}
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,99 @@
+      val namedList = solrClient.request(req)
+      val status = namedList.get("status")
+      if (status != null && status != "0")
+        throw new Exception(s"solr update returned: ${ namedList.asShallowMap().values().toArray().mkString }")
```
I more prefer you using a `Failure` here rather than throwing stuff... Maybe you could use a for-comprehension or `flatMap` here?

jo-pol replied:
Getting lost:

```scala
    for {
      namedList <- Try(solrClient.request(req))
      status = namedList.get("status")
      nonZeroStatus = status != null && status != "0"
    } yield ???
```

rvanheest replied:
```scala
for {
  namedList <- Try(solrClient.request(req))
  status = namedList.get("status")
  _ <- if (status != null && status != "0") Failure(new Exception(s"solr update returned: ${ namedList.asShallowMap().values().toArray().mkString }"))
       else Success(())
} yield ()
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,99 @@
+
+  def deleteBag(bagId: String): Try[FeedBackMessage] = Try {
+    val query = new SolrQuery
+    query.set("q", s"id:$bagId/*")
```
```scala
val query = new SolrQuery {
    set("q", s"id:$bagId/*")
}
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Vault.scala:
```diff
@@ -0,0 +1,42 @@
+    logger.info(s"getting storeNames with $uri")
+    linesFrom(uri).map { line =>
+      val trimmed = line.trim.replace("<", "").replace(">", "")
+      Paths.get(new URI(trimmed).getPath).getFileName.toString
```
I'm surprised you need this extra `URI` here...

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Vault.scala:
```diff
@@ -0,0 +1,42 @@
+import scala.util.Try
+
+trait Vault {
+  this: VaultIO with DebugEnhancedLogging =>
```
use extends for `DebugEnhancedLogging`

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/VaultIO.scala:
```diff
@@ -0,0 +1,60 @@
+import scala.util.Try
+import scala.xml.{ Elem, XML }
+
+trait VaultIO {
```
in general you could do better on error handling in this class. `acquireAndGet` throws the errors. Use `.map(...).tried` instead. _Please note:_ if your `.map` already returns a `Try`, don't forget to do `flatten` after `tried`!

rvanheest commented on src/test/resources/files.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<files xmlns:dcterms="http://purl.org/dc/terms/">
```
next week this file may not be valid anymore... then we need a schema declaration here. Also `dcterms:accessRights` will be changed (@lindareijnhoudt is this actually true?)

rvanheest replied:
would a `<dcterms:title>` have some different behavior? (not sure, just asking)

rvanheest commented on src/test/resources/vault/stores/pdbs/bags/9da0541a-d2c8-432e-8129-979a9830b427/metadata/dataset.xml:
```diff
@@ -0,0 +1,88 @@
+        </dcx-gml:spatial>
+        <dcterms:temporal xsi:type="abr:ABRperiode">PALEOLB</dcterms:temporal>
+        <dcterms:temporal>random text</dcterms:temporal>
+
```
empty lines??? on purpose?

rvanheest commented on src/test/resources/vault/stores/pdbs/bags/9da0541a-d2c8-432e-8129-979a9830b427/metadata/files.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<files xmlns:dcterms="http://purl.org/dc/terms/">
```
see comments on the former `files.xml`

rvanheest commented on src/test/resources/vaultStoreNames/stores:
```diff
@@ -0,0 +1,4 @@
+    <http://localhost:20110/stores/dryad>
```
you really need to use actual names here? maybe use fake names instead...

rvanheest reviewed (changes requested):
fear not, i'm not finished yet.
the tests will be reviewed later

jo-pol committed:
first quick review fixes

jo-pol committed:
collect mixed results

jo-pol committed:
review comments

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -68,35 +68,33 @@ class ApplicationWiring(configuration: Configuration)
-    else {
-      val t = results.collectResults.failed.get
-      Failure(new Exception(s"$stats, another ${ t.getMessage }", t))
+  private implicit class mixedResults(left: Seq[Try[FeedBackMessage]]) {
```
put an extra `extends AnyVal` here for some more optimization: https://docs.scala-lang.org/overviews/core/value-classes.html#extension-methods

jo-pol replied:
causes compiler errors

jo-pol committed:
review comments

jo-pol committed:
collectMixedCreateDocResults with foldLeft

jo-pol committed:
review comments

jo-pol committed:
extends anyval

jo-pol committed:
just an http header for the size (needed before solr opens the stream)

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/FileItem.scala:
```diff
@@ -42,20 +40,18 @@ case class FileItem(bag: Bag, ddm: DDM, xml: Node) extends DebugEnhancedLogging
-  def shouldIndex: Boolean = {
+  val path: String = xml.attribute("filepath").map(_.text).getOrElse("")
+  val accessibleTo: String = ( xml \ "accessRights").map(_.text).mkString match {
+    case s if s.isEmpty => datasetAccessitbleTo
```
1. can't you just match on `""` here?
2. `datasetAccessibleTo`? (without extra `t`)

rvanheest commented on pom.xml:
```diff
@@ -64,6 +64,14 @@
             <artifactId>dans-scala-lib</artifactId>
```
not sure what this version is by default, but in most projects I set it to 1.2 or 1.3

jo-pol replied:
maven dependencies shows 1.2

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/EasyUpdateSolr4filesIndexApp.scala:
```diff
@@ -15,12 +15,20 @@
-class EasyUpdateSolr4filesIndexApp(wiring: ApplicationWiring) extends AutoCloseable {
+import scala.util.{ Success, Try }
 
+class EasyUpdateSolr4filesIndexApp(wiring: ApplicationWiring) extends AutoCloseable
```
Why `AutoClosable`?

jo-pol replied:
from the archetype

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/EasyUpdateSolr4filesIndexApp.scala:
```diff
@@ -15,12 +15,20 @@
-class EasyUpdateSolr4filesIndexApp(wiring: ApplicationWiring) extends AutoCloseable {
+import scala.util.{ Success, Try }
 
+class EasyUpdateSolr4filesIndexApp(wiring: ApplicationWiring) extends AutoCloseable
```
Given that this class only does method forwarding and nothing else (right?), doesn't it make sense to remove this class and either use the `ApplicationWiring` instead of this in higher classes, or rename `ApplicationWiring` to `EasyUpdateSorl4filesIndex`?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/EasyUpdateSolr4filesIndexApp.scala:
```diff
@@ -15,12 +15,20 @@
+
+  def delete(bagId: String): Try[String] = wiring.delete(bagId)
 
   def init(): Try[Unit] = {
```
If you ain't need it, remove it

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Bag.scala:
```diff
@@ -0,0 +1,62 @@
+    vaultIO.fileURL(storeName, bagId, path)
+  }
+
+  private val fileShas: FileToShaMap = Try {
```
note that this is a `val`, which means it gets evaluated on constructor time. Don't you want lazy evaluation here? Not sure...

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,116 @@
+          case map if map.isEmpty => Seq(("dataset_subject", n.text))
+          case abrMap => Seq(
+            ("dataset_abr_subject", n.text), // TODO https://github.com/DANS-KNAW/easy-dtap/pull/131#pullrequestreview-60582037
+            ("dataset_subject", abrMap.get.getOrElse(n.text, ""))
```
This `abrMap.get` fails if `abrMap` is a `None`.
Also I'm not sure if the pattern match is correct: in case 1 you match on a map where the map is empty. However, the type of this map is `Option[Map[X, Y]]`, so you effectively only check on whether the map value is a `None`.

Shouldn't it be something like this:

```scala
getAbrMap(n) match {
  case Some(map) if map.isEmpty => ...
  case Some(map) => ...
  case None => ...
}
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,116 @@
+      new URL(xsdURL).openStream()
+    ).acquireAndGet(XML.load) // TODO error handling
+    (xmlDoc \ "simpleType")
+      .map(n => (n.attribute("name").head.text, findKeyValuePairs(n)))
```
`n.attribute("name").head.text`

what if `n.attribute("name")` returns an empty list? Then `.head` throws a `NoSuchElementException`! Please fix!

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,85 @@
+      addContentStream(new ContentStreamBase.URLStream(item.bag.fileUrl(item.path)))
+      setParam("literal.id", solrDocId)
+      for ((key, value) <-
+           ("file_size", size.toString) +:
```
so, all you do in this for loop is iterate over the `(key, value)` pairs, check whether the value is not an empty `String` and then call `setParam`. However, outside the for-loop you already do this for `("literal.id", solrDocId)`. Why don't you handle `("file_size", size.toString)` there as well? You are certain that `size` isn't an empty `String` since you converted it from a `Long`. So why first add it to the other aggregated collection?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,85 @@
+  private def executeUpdate(req: ContentStreamUpdateRequest): Try[Unit] = {
+    for {
+      namedList <- Try(solrClient.request(req))
+      status = namedList.get("status")
```
you can also do this like `_ <- Option(namedList.get("status")).withFilter("0" !=).map(Failure(...)).getOrElse(Success(())`, which merges these two lines and is more 'Scala-esc'

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Vault.scala:
```diff
@@ -0,0 +1,62 @@
+  def getBagIds(storeName: String): Try[Seq[String]] = Try {
+    val storeURI = vaultBaseUri.resolve(s"stores/$storeName/bags")
+    logger.info(s"getting bag ids with $storeURI")
+    linesFrom(storeURI).map { _.trim }
```
just a matter of style: `.map(_.trim)`

`()` instead of `{}`

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Vault.scala:
```diff
@@ -0,0 +1,62 @@
+        logger.warn(s"getSize($url) ${ response.statusLine }, details: ${ response.body }")
+        -1L
+      case response =>
+        Try(response.headers("content-length").toLong).recoverWith {
```
Mmm, I got a feeling you require some other operators. `Try(...)` might be a `Success` or a `Failure`, but with this `recoverWith` you transform every `Failure` into a `Success(-1L)`. However, `recoverWith` still returns a `Try`, such that you use a `getOrElse(-1L)` to get the real value, even though this last operator will never get called. Instead you might want to do: `Try(...).doIfFailure { case e => logger.warn(..., e) }.getOrElse(-1L)`.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -15,12 +15,41 @@
+    }
+  }
+
   implicit class TryExtensions2[T](val t: Try[T]) extends AnyVal {
```
If you don't use it, please remove it!

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -0,0 +1,114 @@
+
+  "update" should "call the stubbed solrClient.request" in {
+    val result = new MockedAndStubbedWiring().update(store, uuid)
+    inside(result) {
```
inline `result`
same on other places...
rule of thumb: if inlining keeps it on one line, inline it

jo-pol committed:
dropped redundant code

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -0,0 +1,114 @@
+      .getParams.getMap.asScala
+      .toSeq.sortBy { case (k, _) => k }
+      .toMap
+      .count { case (_, values) =>
```
if `values` is an empty `Array[String]`, you get an error here... You can do a pattern match in the lambda:
```scala
case (_, Array()) => ???
case (_, Array(head, _*)) => head.toLowerCase...
```

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -0,0 +1,114 @@
+      override def update(store: String, uuid: String) = Failure(new Exception("stubbed ApplicationWiring.update"))
+    }.initSingleStore(store)
+    inside(result) { case Failure(e) =>
+      e.getMessage shouldBe "Tried to update 5 bags, 5 exceptions occurred."
```
`e should have message ...`

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -0,0 +1,114 @@
+      // vaultBagIds/bags can't be a file and directory so we need a stub, a failure proofs its called
+      override def update(store: String, uuid: String) = Failure(new Exception("stubbed ApplicationWiring.update"))
+    }.initSingleStore(store)
+    inside(result) { case Failure(e) =>
```
can't you do a further pattern match on `e`, for example:

`case Failure(MyCustomException(msg, CompositeException(headCause :: _))) =>`

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -0,0 +1,114 @@
+      // vaultStoreNames/stores can't be a file and directory so we need a stub, a failure proofs its called
+      override def initSingleStore(store: String) = Failure(new Exception("stubbed ApplicationWiring.initSingleStore"))
+    }.initAllStores()
+    inside(result) { case Failure(e) =>
```
same comment as above about further pattern matching `e`

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -0,0 +1,114 @@
+
+  private def createConfig(testDir: String) = {
+    val vaultPath = URLEncoder.encode(Paths.get(s"src/test/resources/$testDir").toAbsolutePath.toString, "UTF8")
+    val properties = new PropertiesConfiguration()
```
use shorter syntax: 

```scala
new PropertiesConfiguration() {
  addProperty(...)
  ...
}

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/CustomMatchers.scala:
```diff
@@ -17,20 +17,21 @@ package nl.knaw.dans.easy.solr4files
 import java.io.File
 
-import scala.io.Source
+import org.scalatest.matchers.{ MatchResult, Matcher }
```
I assume formatting issues in `CustomMatchers`, `DebugConfigSpec` and `ReadmeSpec` come from the archetype? Maybe also fix these over there.

jo-pol replied:
Formatting can trigger codacy warnings you might not want to trigger without solving it in the archetype and all other projects with the same code snippets.

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/components/DDMSpec.scala:
```diff
@@ -0,0 +1,94 @@
+class DDMSpec extends FlatSpec with Matchers {
+
+  "solrLiteral" should "return proper values" in {
+    assume(canConnectToEasySchemas)
```
instead of assuming that we can connect to easy-schema, maybe we should fail if we can't connect. In that case something is terribly wrong with easy01 and a failed test will show up more easily than an ignored test

jo-pol replied:
Then you can't build when offline while traveling.

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/components/DDMSpec.scala:
```diff
@@ -0,0 +1,94 @@
+  "solrLiteral" should "return proper values" in {
+    assume(canConnectToEasySchemas)
+    val ddm = new DDM(resource.managed(new FileInputStream(
+      "src/test/resources/vault/stores/pdbs/bags/9da0541a-d2c8-432e-8129-979a9830b427/metadata/dataset.xml"
```
preferably use the copy in `target` for this using `getClass.getResource()`. See for example

https://github.com/DANS-KNAW/easy-split-multi-deposit/blob/69957d9f2244092f88d49eb903c6cb0bc781ee32/src/test/scala/nl.knaw.dans.easy.multideposit/BlackBoxSpec.scala#L62-L63

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/components/DDMSpec.scala:
```diff
@@ -0,0 +1,94 @@
+      ("dataset_title", "Trip to Centaur asteroid")
+    )
+    // verbose check to figure out which one is wrong in case of problems
+    expected.foreach(literals should contain(_))
```
`expected should contain theSameElementsAs literals`?
or maybe `theSameElementsInOrderAs`?

jo-pol replied:
How can I clarify the comment?

rvanheest reviewed (changes requested):
Looks great already. I now also reviewed the tests...

jo-pol committed:
moved vaultIO to implicit class RichURL

jo-pol committed:
review comments

jo-pol committed:
implicit class RichURL used for all IO, lazyness

jo-pol committed:
lazyness revisited, statics

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Bag.scala:
```diff
@@ -17,31 +17,37 @@ package nl.knaw.dans.easy.solr4files.components
+      lines <- url.readLines
+      wantedLines = lines.filter(_.trim.startsWith(key))
+      values = wantedLines.map(_.trim.replace(key, "").trim.replace(":", "").trim)
+      value <- Try(values.head)
```
I keep repeating this point: you can't do a `head`! It's dangerous, even though you wrapped it in a `Try` (which in itself isn't necessary here either). What happens if `values` is empty?

Instead either do a pattern match, or in this case even better: use the failsafe `headOption`. Then you have to deal with the option, but with that you can specify a custom exception...

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -87,23 +100,14 @@ object DDM {
+      .map(n => (n.attribute("name").map(_.text).getOrElse(""), findKeyValuePairs(n)))
       .toMap
-  }
+  }.getOrElse(Map[String, Map[String, String]]())
```
replace with `Map.empty`. The compiler already knows about the specific types...

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -66,12 +67,15 @@ trait Solr extends DebugEnhancedLogging {
-      _ <- if (status == null || status == "0") Success(())
-           else Failure(new Exception(s"solr update returned: ${ namedList.asShallowMap().values().toArray().mkString }"))
+      status = Option(namedList.get("status")).withFilter("0" !=)
+      _ <- status.map(_ => statusError(namedList)).getOrElse(Success(()))
```
why not merge these two lines?

jo-pol replied:
Too long to read, I'd get lost.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Vault.scala:
```diff
@@ -16,47 +16,40 @@
+  }
+
+  @throws // (NullPointerException,IllegalArgumentException)
+  def fileURL(storeName: String, bagId: String, file: String): URL = {
```
Why use `@throws` here, rather than returning a `Try[URL]`?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/VaultIO.scala:
```diff
@@ -1,60 +0,0 @@
-import scala.util.Try
-import scala.xml.{ Elem, XML }
-
-trait VaultIO {
```
what happened to the functions in this class?

rvanheest reviewed (commented):
some comments on your commits from yesterday.

jo-pol committed:
not just check the http-code, also catch exceptions

jo-pol committed:
replaced URL.openstream

jo-pol committed:
replaced throws with try

jo-pol committed:
review comments

jo-pol committed:
code formatting

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -60,26 +59,38 @@ package object solr4files extends DebugEnhancedLogging {
-    def loadXml: Try[Elem] = Try {
-      resource.managed(left.openStream()).acquireAndGet(XML.load) // TODO replace
+    def loadXml: Try[Elem] = {
+      getContent.flatMap(s => Try(XML.loadString(s)))
```
since `getContent` is already a `Try`, you don't need to wrap `XML.loadString` in a `Try` anymore. Any kind of error handling will be dealt with by `map`.

`getContent.map(XML.loadString)`

rvanheest reviewed (commented):
Starts looking real good!

rvanheest commented:
@jo-pol maybe it's a good idea merge this PR before adding the service components? This PR is already really big and I'm afraid we loose the overview when you start adding even more stuff.

jo-pol committed:
custom exceptions (breaks tests)

jo-pol commented on src/test/scala/nl.knaw.dans.easy.solr4files/ApplicationWiringSpec.scala:
```diff
@@ -99,7 +99,8 @@ class ApplicationWiringSpec extends TestSupportFixture {
     }.initAllStores()
-    inside(result) { case Failure(Exception(msg, CompositeException(headCause :: _))) =>
+    inside(result) {
+      case Failure(WrappedCompositeException(msg, CompositeException(headCause :: _))) =>
```
This test breaks because it checks the msg passed into the exception, I want to test against the created message.

rvanheest replied:
which message do you mean (don't know what the created message is...)? The one from `CompositeException`? Then you add `e @ ` before `CompositeException`, right?

jo-pol committed:
fixed unit tests

jo-pol committed:
another custom exception

jo-pol committed:
another custom exception and further error handling

jo-pol committed:
another map.empty and should have

PaulBoon commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -15,15 +15,79 @@
+
+  def delete(bagId: String): Try[FeedBackMessage] = for {
+    _ <- deleteBag(bagId)
+    _ <- commit()
```
What happens if some of the files have been deleted, but there is an exception and we do not do the commit. Then those deletes are not committed, but will that be done with the next command if we run as a service..? Maybe we should try a commit if something fails, or we need to understand what happens if we do not commit.

PaulBoon commented on src/main/scala/nl.knaw.dans.easy.solr4files/Command.scala:
```diff
@@ -44,6 +41,15 @@ object Command extends App with DebugEnhancedLogging {
+        case init @ commandLine.init => init.bagStore.toOption
+          .map(app.initSingleStore)
+          .getOrElse(app.initAllStores())
+        case commandLine.runService => ???
```
Don't forget this one, there is no checkbox in the PR and no TODO either.

rvanheest replied:
If you would check your email more often, you would have seen this comment:

https://github.com/DANS-KNAW/easy-update-solr4files-index/pull/1#issuecomment-328852321

PaulBoon commented on src/main/scala/nl.knaw.dans.easy.solr4files/CommandLineOptions.scala:
```diff
@@ -15,18 +15,24 @@
     s"""
-       |  $printedName (synopsis of command line parameters)
-       |  $printedName (... possibly multiple lines for subcommands)""".stripMargin
+       |  $printedName {add|delete|update} [-s <bag-store>] <uuid>
```
We don't support the 'add' anymore, but I see it here.

PaulBoon commented on src/main/scala/nl.knaw.dans.easy.solr4files/CommandLineOptions.scala:
```diff
@@ -39,14 +45,39 @@ class CommandLineOptions(args: Array[String], configuration: Configuration) exte
+    val bagUuid: ScallopOption[UUID] = trailArg(name = "bag-uuid", required = true)
+    footer(SUBCOMMAND_SEPARATOR)
+  }
+  val init = new Subcommand("init") {
```
Do we need this, you can do delete and update?

jo-pol replied:
Update does a single bag, this one all bags.

PaulBoon commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -0,0 +1,120 @@
+  }.getOrElse(Map.empty)
+
+  private def findKeyValuePairs(table: Node): Map[String, String] = {
+    (table \\ "enumeration")
```
You put all codes in one map, so both complex and period types. Now the only overlap seems to be 'XXX', being 'Onbekend', but it might be safer to use two maps and make it simpler to treat them differently later on.

jo-pol replied:
the caller takes care of the different tables

PaulBoon commented on README.md:
```diff
@@ -26,14 +26,36 @@ ARGUMENTS
 --------
 
-    easy-update-solr4files-index -o value
+    easy-update-solr4files-index add 9da0541a-d2c8-432e-8129-979a9830b427
```
Should be 'update' instead of 'add'?

jo-pol committed:
add in readme, commit even if we had some errors

jo-pol committed:
license header

jo-pol committed:
readme

jo-pol committed:
in sync with https://github.com/DANS-KNAW/easy-dtap/pull/131 8ceeb6ca0

jo-pol committed:
fail fast, commit everything

jo-pol committed:
white space for one liners

adapted from https://stackoverflow.com/questions/39001421/scala-what-is-the-easiest-way-to-get-all-leaf-nodes-and-their-paths-in-an-xml

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -58,36 +57,71 @@ class ApplicationWiring(configuration: Configuration)
+  }.recoverWith {
+    case t: SolrCommitException => Failure(t)
+    case t =>
+      commit()
```
if this `commit()` returns a `Failure` as well, you'll loose it...

```scala
commit().flatMap(_ => Failure(t))
  .getOrRecover(t2 => new CompositeException(t2, t))
```

jo-pol replied:
tried variations but don't get it compiling.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -58,36 +57,71 @@ class ApplicationWiring(configuration: Configuration)
+
+    val stats = s"Updated ${ bagIds.size } bags for $storeName"
+
+    if (thrown.isEmpty) Success(stats)
```
why not do a pattern match on `(throw, results)`? That's why you got this tuple, after all... Something like:

```scala
lazy val stats = s"Updated ${ bagIds.size } bags for $storeName"

bagIds.toStream.map(update(storeName, _)).takeUntilFailure match {
  case (None, _) => Success(stats)
  case (Some(t), Seq()) => Failure(t)
  case (Some(t), _) => Failure(AnotherFailedException(stats, t))
}
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -58,36 +57,71 @@ class ApplicationWiring(configuration: Configuration)
-  private def updateFiles(bag: Bag, ddm: DDM, filesXML: Elem) = {
-    (filesXML \ "file")
+  private def updateFiles(bag: Bag, ddm: DDM, filesXML: Elem): Try[FeedBackMessage] = {
+    val (thrown, results) = (filesXML \ "file")
```
same here?

jo-pol replied:
Would like it, but in this third case the partitioned results make it more complicated.

rvanheest replied:
We'll look at it on Monday to see what we can do. Honestly I didn't really look at this one as it looked the same as the one before...

rvanheest replied:
```scala
val (thrown, results) = (filesXML \ "file").<etc.>

// note these lazy keywords!
lazy val (withContent, justMetadata) = results.partition(_.isInstanceOf[SubmittedWithContent])
lazy val stats = s"bag ${ bag.bagId }: updated ${ withContent.size } files with content, ${ justMetadata.size } without content"

(thrown, results) match {
  case (Some(e), Seq()) => Failure(e)
  case (Some(e), rs) =>
    for (c <- withContent)
      logger.info(c.solrId)
    Failure(AnotherFailedException(stats, e))
  case (None, _) => Success(stats)
}
```

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/DDM.scala:
```diff
@@ -85,6 +96,22 @@ object DDM {
+  def spacedText(n: Node): String = {
+    val s = mutable.ListBuffer[String]()
+
+    def strings(n: Seq[Node]): Unit =
```
this looks and feels like recursion. You can easily turn this into tail recursion. Hint: pattern matching on `n`.

rvanheest replied:
I was thinking about this some more, if you don't mind...

```scala
def spacedText(ns: Seq[Node]): String = {

  @tailrec
  def internal(todo: Seq[Node] = ns, result: List[String] = List.empty): List[String] = {
    todo match {
      case Seq() => result
      case Seq(h, t@_*) if h.child.isEmpty => internal(t, h.text :: result)
      case Seq(h, t@_*) => internal(h.child ++ t, result)
    }
  }

  internal().reverse.mkString(" ")
}
```

rvanheest replied:
brief explanation:
* I add the `h.text` to the front of the list (prepend), since `::` is an O(1) operation. This in contrast to adding to the end of the list (append), which I'm not sure about, but I believe it's like O(n).
* To make sure the desired order is correct, I have to do a `reverse` at the end. Note again that it is way more efficient to do a single `reverse` operation at the end than do an append for every node in the tree!

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -16,49 +16,57 @@
 
-  case class WrappedCompositeException(msg: String, cause: CompositeException )
-    extends Exception(s"$msg ${cause.getMessage()}", cause)
+  case class WrappedCompositeException(msg: String, cause: CompositeException)
```
I'm not yet a fan of this name. Maybe you could think of a better name or have better namings for the specific use cases.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -16,49 +16,57 @@
+    extends Exception(s"$msg ${ cause.getMessage() }", cause)
 
-  case class HttpStatusException(msg: String, response: HttpResponse[String] )
+  case class AnotherFailedException(msg: String, cause: Throwable)
```
Same here, I'm not a fan of this name. I can't make anything of this. The other custom exceptions below such as `HttpStatusException`, `SolrStatusException` or `SolrCommitException` have way clearer names.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -16,49 +16,57 @@
+      val it = left.iterator
+      while (it.hasNext) {
+        val x = it.next()
+        if (x.isFailure) return (Some(x.failed.get), b.toList)
```
do a pattern match here:

```scala
it.next() match {
  case Success(x) => b += x
  case Failure(t) => return (Some(t), b.toList)
}
```

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/tmp.sc:
```diff
@@ -0,0 +1,33 @@
+import scala.collection.mutable
```
Remove this from your commit. And add `*.sc` to your `.gitignore`.

rvanheest reviewed (changes requested):
Here are some more comments on your latest commits.

jo-pol committed:
tested with a real one-liner

rvanheest commented on src/test/scala/nl.knaw.dans.easy.solr4files/components/DDMSpec.scala:
```diff
@@ -63,18 +63,11 @@ class DDMSpec extends TestSupportFixture {
-            </dcx-dai:organization>
-          </dcx-dai:author>
-        </dcx-dai:creatorDetails>
+        <dcx-dai:creatorDetails><dcx-dai:author><dcx-dai:titles>Captain</dcx-dai:titles><dcx-dai:initials>J.T.</dcx-dai:initials><dcx-dai:surname>Kirk</dcx-dai:surname><dcx-dai:organization><dcx-dai:name xml:lang="en">United Federation of Planets</dcx-dai:name></dcx-dai:organization></dcx-dai:author></dcx-dai:creatorDetails>
```
what happened to your formatting?

jo-pol committed:
dropped redundant recursion

jo-pol committed:
added identifiers

jo-pol committed:
dct temporal coverage from https://github.com/DANS-KNAW/easy-schema/blob/master/src/main/assembly/dist/docs/examples/ddm/example2.xml#L122

jo-pol committed:
spatial coverage as far as descriptive; implicit class for XML.Node

jo-pol committed:
simple coverage, formatting and extracted methods

jo-pol committed:
boolean methods to implicit class RichNode

jo-pol committed:
boolean methods to implicit class RichNode

jo-pol committed:
drop work sheet

jo-pol committed:
ignore worksheets

jo-pol committed:
review: tail recursion

jo-pol committed:
if-else to pattern matching

jo-pol committed:
exception rename, still not happy

jo-pol committed:
another pattern match

jo-pol commented on src/main/scala/nl.knaw.dans.easy.solr4files/ApplicationWiring.scala:
```diff
@@ -82,37 +82,37 @@ class ApplicationWiring(configuration: Configuration)
+          results // FilesSubmittedWithJustMetadata logged warnings
+            .withFilter(_.isInstanceOf[FilesSubmittedWithContent])
+            .foreach(x => logger.info(x.toString))
+          Failure(MixedResultsException(statsPrefix, results, t))
```
Disadvantage of this solution: in the last case the results are walked twice.

rvanheest replied:
why?

jo-pol replied:
First by the withFilter above, then within MixedResultsException to assemble the stats. Anyway it only happens when a file fails after some succeeded. On the other hand, could very large number of files problem?

jo-pol committed:
error handling (without further unit tests)

jo-pol committed:
code formatting

jo-pol committed:
eliminated SomeSucceededException, now using MixedResultsException

jo-pol committed:
fixed license header

jo-pol committed:
fixed some warnings

jo-pol committed:
tailrec eliminates codacy warning, moved the method

jo-pol committed:
comment about logging lied

jo-pol committed:
takeUntilFailure returned a tuple, now Try[Seq[T]] like collectResults

jo-pol committed:
eliminate compiler warning, plural class names to single

jo-pol committed:
fixed manual tests: all stores, feedback messages

rvanheest commented:
* see easy-pid-generator for a correct example of the default `logback.xml` logger patterns.
* ansible: make sure you have the correct requirements versions
* ansible: `vagrant.yml` is missing the roles for installing an easy-bag-store and the solr service you're using in this module

jo-pol committed:
fixed unit tests

jo-pol committed:
whoops, previous commit was incomplete

jo-pol committed:
debug logging with solr request contents

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:
```diff
@@ -0,0 +1,103 @@
+        addContentStream(new ContentStreamBase.URLStream(fileUrl))
+        setParam("literal.id", solrDocId)
+        setParam("literal.easy_file_size", size.toString)
+        for (
```
replace this for-loop with

```scala
for ((key, value) <- item.bag.solrLiterals ++ item.ddm.solrLiterals ++ item.solrLiterals
     if value.trim.nonEmpty) {
  setParam(s"literal.easy_$key", value.replaceAll("\\s+", " ").trim)
}
```

This will on byte-code level transform the `if ...` into a `withFilter` and the body of the for-loop into a `foreach`.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -15,20 +15,144 @@
+    extends Exception(thrown.getMessage, thrown)
+
+  implicit class RichSeq[T](val left: Seq[T]) extends AnyVal {
+    def stats: String = {
```
you only use this method in one place now. Maybe inline it?

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -15,20 +15,144 @@
+
+  implicit class RichTryStream[T](val left: Seq[Try[T]]) extends AnyVal {
+
+    /** Typical usage: val (thrown, results) = ...toStream.map(TrySomething).takeUntilFailure */
```
this is still the old API...

rvanheest commented on src/main/scala/nl.knaw.dans.easy.solr4files/package.scala:
```diff
@@ -15,20 +15,144 @@
+        else it.next() match {
+          case Success(y) => inner(y :: result)
+          case Failure(t) =>
+            val seq: Seq[T] = result.reverse
```
inline this line if possible (maybe it's actually needed because of the explicit type declaration???)

jo-pol committed:
logback of archetype was behind conventions

jo-pol committed:
formatting

jo-pol committed:
restored listBuffer in tailrec

rvanheest reviewed (approved):
Looks great!

jo-pol committed:
listbuffer for the other tailrec

jo-pol committed:
listbuffer in both tailrecs outside the inner method

jo-pol committed:
fix unicode warning by codacy

jo-pol committed:
description details in README
